usecase/interactor/pokeinteractor.go

```diff
@@ -0,0 +1,149 @@
+package interactor
+
+import (
+	"errors"
+	"fmt"
+	"github.com/ToteEmmanuel/academy-go-q12021/domain/model"
+	"github.com/ToteEmmanuel/academy-go-q12021/usecase/presenter"
+	"github.com/ToteEmmanuel/academy-go-q12021/usecase/repository"
+	"github.com/go-resty/resty/v2"
+	"sync"
+)
+
+type pokeInteractor struct {
+	PokeRepo      repository.PokeRepository
+	PokePresenter presenter.PokePresenter
+	RestClient    *resty.Client
+}
+
+type PokeInteractor interface {
+	Get(id int32) (*model.Pokemon, error)
+	GetAll() ([]*model.Pokemon, error)
+	CatchOne(int32) (*model.Pokemon, error)
+}
+
+func NewPokeInteractor(r repository.PokeRepository, p presenter.PokePresenter, client *resty.Client) PokeInteractor {
+	return &pokeInteractor{
+		PokeRepo:      r,
+		PokePresenter: p,
+		RestClient:    client,
+	}
+}
+
+func (pI *pokeInteractor) Get(id int32) (*model.Pokemon, error) {
+	p, err := pI.PokeRepo.FindById(id)
+	if err != nil {
+		return nil, err
+	}
+	return pI.PokePresenter.ResponsePokemon(p), nil
+}
+
+func (pI *pokeInteractor) GetAll() ([]*model.Pokemon, error) {
+	p, err := pI.PokeRepo.FindAll()
+	if err != nil {
+		return nil, err
+	}
+	return pI.PokePresenter.ResponsePokemons(p), nil
+}
+
+func (pI *pokeInteractor) CatchOne(id int32) (*model.Pokemon, error) {
+	p, err := pI.lookForPokemon(id)
+	if err != nil {
+		return nil, err
+	}
+	return pI.PokePresenter.ResponsePokemon(p), nil
+}
+
+func (pI *pokeInteractor) lookForPokemon(id int32) (*model.Pokemon, error) {
+	var waitgroup sync.WaitGroup
+	waitgroup.Add(2)
+	catchedPokemon := &model.Pokemon{
+		Id:         id,
+		Species:    "",
+		Sprite:     "",
+		FlavorText: "",
+		Types:      nil,
+	}
+	var err error
+	err = fetchBase(catchedPokemon, pI.RestClient)
+	if err != nil {
+		fmt.Println(err)
+	}
+	go func() {
+		err = fetchSpecies(catchedPokemon, pI.RestClient)
+		waitgroup.Done()
+		fmt.Println("Fetched Species...")
+	}()
+	go func() {
+		err = fetchBase(catchedPokemon, pI.RestClient)
+		waitgroup.Done()
+		fmt.Println("Fetched Base...")
+	}()
+	waitgroup.Wait()
```

Comment:
You would probably like to think of a way of handling both errors you're receiving from your go routines, if both of your APIs fail, then you will only know about one failure instead of both. There's probably a way of doing this with channels

Reply:
followed the advice and used a channel to handle the errors.